pkg/controller/infrastructure/infraflow/ensure_utils.go

```diff
@@ -296,3 +299,145 @@ func isUserRouter(config *gcp.InfrastructureConfig) bool {
 func isUserVPC(config *gcp.InfrastructureConfig) bool {
 	return config.Networks.VPC != nil && len(config.Networks.VPC.Name) > 0
 }
+
+// isSimilar return whether to slices are similar using reflect.DeepEqual
+//
+// Similar in this context means they contain the same elements, not necessarily in the same order.
+// In essence, this is set equality using DeepEqual.
+func isSimilar[T any](s1, s2 []T) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	for _, v := range s1 {
+		if !slices.ContainsFunc(s2, func(e T) bool { return reflect.DeepEqual(v, e) }) {
+			return false
+		}
+	}
+	return true
+}
+
+// firewallUpdate returns a firewall rule that reflects the differences from 'left' towards 'right'.
+//
+// The intended use for the returned firewall rule is to serve as a minimal patch when updating
+// firewall rules on gcp.
+func firewallUpdate(left, right *compute.Firewall) (*compute.Firewall, error) {
+	if left == right {
+		// both identical, nothing to do
+		return nil, nil
+	}
+	if left == nil && right != nil {
+		// everything is right, return right
+		return right, nil
+	}
+	if right == nil && left != nil {
+		// strictly speaking, the update here would be an empty Firewall-object with its
+		// ForceSendFields set to all fields. In the intended context of this function,
+		// just return nil signalling that there is nothing to do
+		return nil, nil
+	}
+
+	// The basic idea is to construct a map holding all the differences from left to right and
+	// then use Golang's json-marshalling to create an actual Firewall-instance with the
+	// corresponding attributes set to the expected values.
+	// If only there was a way to supply function args from maps, like foo(**kwargs) :(
+	diff := map[string]interface{}{}
+
+	// Note: CreationTimestamp is ignored, everything else is checked.
+
+	if len(right.Allowed) != 0 {
+		// n.b.: This replaces the entries on the left, it does not merge (same behaviour as Terraform)
+		if len(left.Allowed) == 0 || !isSimilar(left.Allowed, right.Allowed) {
+			diff["allowed"] = right.Allowed
+		}
+	}
+	if len(right.Denied) != 0 {
+		if len(left.Denied) == 0 || !isSimilar(left.Denied, right.Denied) {
+			diff["denied"] = right.Denied
+		}
+	}
+	if right.Description != "" && left.Description != right.Description {
+		diff["description"] = right.Description
+	}
```

Comment:
careful here, the description is immutable. We would have to delete and then recreate the rule.